### PR TITLE
added rng_guide

### PR DESCRIPTION
**Pull request by chiamp**
Added RNG guide. View preview [here](https://flax--3497.org.readthedocs.build/en/3497/guides/flax_fundamentals/rng_guide.html).

Follow-up questions:
* should we add a section to show how to split RNG's using `pmap`?
* should we add an additional example using `nn.scan` for sequential models and modifying the `variable_broadcast` arg?
* should the rng guide be placed under the "Flax fundamentals" section, or somewhere else like "Developer notes"?

### CODE REVIEW

**Comment by cgarciae** [comment]
> should we add a section to show how to split RNG's using pmap?

I think the `vmap` example is close enough.

> should we add an additional example using nn.scan for sequential models and modifying the variable_broadcast arg?

I don't think its necessary, shouldn't change the rng-related behavior.

> should the rng guide be placed under the "Flax fundamentals" section, or somewhere else like "Developer notes"?

I'd vote for "Flax fundamentals".

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+## Setup
+
++++
+
+Import some necessary dependencies.
+
+**Note:** This guide uses the `--xla_force_host_platform_device_count=8` flag to emulate multiple devices in a CPU environment in a Google Colab/Jupyter Notebook. You don’t need this if you are already using a multi-device TPU environment.
+
+```{code-cell}
+!pip install -q flax
```
Comment (cgarciae) [inline docs/guides/flax_fundamentals/rng_guide.md:28]: we need to add the `skip-execution` tag to this cell

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+```
+
+### How `self.make_rng` works under the hood
+
++++
+
+When `self.make_rng` is called, the following happens:
+* The following data is collected:
+  * the name of the module making the `self.make_rng` call (note: the top-level root module has no name)
+  * the name of all of its parent modules, all the way up until the top-level root module
```
Comment (cgarciae) [inline docs/guides/flax_fundamentals/rng_guide.md:121]:
```suggestion
  * the path of the Module as given by `self.scope.path` (the top-level root module has an empty path `()`)
```

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 3]
```diff
@@ -0,0 +1,650 @@
...
+
+```{code-cell}
+:outputId: 21e623da-91d2-4bbc-e167-d3b257344124
+
+flax.config.update('flax_fix_rng_separator', True)
+out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
+out1 == out2 # different output
+```
+
+## Managing RNG's on multiple devices
```
Comment (cgarciae) [inline docs/guides/flax_fundamentals/rng_guide.md:424]: Wondering if we should recommend `shard_map` instead?

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 4]
```diff
@@ -0,0 +1,650 @@
...
+BatchDense = nn.vmap(
+    nn.Dense,
+    in_axes=0, out_axes=0,
+    variable_axes={'params': None},
+    split_rngs={'params': False})
+
+BatchDense(2).init(jax.random.key(0), x)
+```
+
+By denoting `variable_axes={'params': 0}'`, we can generate parameters for each batch input. However the parameter values generated are all identical to each other:
```
Comment (cgarciae) [inline docs/guides/flax_fundamentals/rng_guide.md:536]:
```suggestion
By denoting `variable_axes={'params': 0}'`, vectorize the `params` Arrays on the first axis. However the parameter values generated are all identical to each other:
```

File: docs/guides/flax_fundamentals/rng_guide.md (hunk 3 above)
Comment (chiamp) [reply docs/guides/flax_fundamentals/rng_guide.md:424]: what about having both an example for `jit` and and for `shard_map`?

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+:outputId: 1f711ced-a97a-49e5-e92c-15ee6d57f93a
+
+variables = rng_module_two_streams.init({'rng_stream1': jax.random.key(0), 'rng_stream2': jax.random.key(0)})
+```
+
+### How `self.make_rng` works under the hood
+
++++
+
+When `self.make_rng` is called, the following happens:
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:120]: Suggestion:

```suggestion
When `self.make_rng` (`flax.linen.Module.make_rng`) is called, the following happens:
```

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+  print(jax.random.fold_in(stream_seed, jnp.uint32(hash_int)))
+```
+
+```{code-cell}
+:outputId: 89d7cce4-da43-4393-e07c-73f2c53cefe4
+
+variables = rng_module.init({'rng_stream': jax.random.key(0)})
+```
+
+### Submodules
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:164]:
```suggestion
### Sub-`Module`s
```

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+    elif isinstance(x, int):
+      m.update(x.to_bytes((x.bit_length() + 7) // 8, byteorder='big'))
+    else:
+      raise ValueError(f'Expected int or string, got: {x}')
+  d = m.digest()
+  hash_int = int.from_bytes(d[:4], byteorder='big')
+  return hash_int
+```
+
+Following the instructions above, we can manually reproduce the PRNG keys generated from `self.make_rng`.
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:147]:
```suggestion
Following the instructions above, you can manually reproduce the PRNG keys generated from `self.make_rng` as follows:
```

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+:outputId: 89d7cce4-da43-4393-e07c-73f2c53cefe4
+
+variables = rng_module.init({'rng_stream': jax.random.key(0)})
+```
+
+### Submodules
+
++++
+
+Let's take a look at how `self.make_rng` behaves with submodules.
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:168]:
```suggestion
This section explores how `self.make_rng` (`flax.linen.Module.make_rng`) behaves with sub-`Module`s.

For example:
```

File: docs/guides/flax_fundamentals/rng_guide.md (hunk 3 above)
Comment (8bitmp3) [reply docs/guides/flax_fundamentals/rng_guide.md:425]: +1 to both `(p)jit` and `shard_map`

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+    module_name = initial_data[-1]
+  else:
+    module_name = 'RNGModule'
+  for call_count in (1, 2):
+    hash_int = produce_hash(data=initial_data+(call_count,))
+    rng_key = jax.random.fold_in(stream_seed, jnp.uint32(hash_int))
+    print(f"{module_name}, count {call_count}: {rng_key}")
+```
+
+If the same submodule class is used multiple times, we simply increment the suffix of the submodule name accordingly; i.e. `RNGSubModule_0`, `RNGSubModule_1`, etc.
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:218]: Nit:

```suggestion
If the same sub-`Module` class is used multiple times, you can increment the suffix of the sub-`Module` name accordingly. For example: `RNGSubModule_0`, `RNGSubModule_1`, and so on.
```

**Comment by chiamp** [comment]
cc: @froystig

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 10]
```diff
@@ -0,0 +1,774 @@
...
+    return Leaf(name=self.leaf_name)(x)
+
+class Model(nn.Module):
+  @nn.compact
+  def __call__(self, x):
+    return (Node(name="ab", leaf_name="cdef")(x),
+            Node(name="abc", leaf_name="def")(x),
+    )
+
+out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
```
Comment (froystig) [inline docs/guides/flax_fundamentals/rng_guide.md:450]:
```suggestion
out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.key(33)})
```
Would this work?

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 11]
```diff
@@ -0,0 +1,774 @@
...
+print(produce_hash(data=('ABC', 1)))
+```
+
+To avoid this edge case, users can flip the `flax_fix_rng_separator` [configuration flag](https://flax.readthedocs.io/en/latest/api_reference/flax.config.html#flax.configurations.Config.flax_fix_rng_separator) to `True`.
+
+```{code-cell}
+:outputId: d7905467-1b77-4f7f-f64c-607d4c97e1c8
+
+flax.config.update('flax_fix_rng_separator', True)
+out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.PRNGKey(33)})
```
Comment (froystig) [inline docs/guides/flax_fundamentals/rng_guide.md:471]:
```suggestion
out1, out2 = Model().apply({}, 0, rngs={"rng": jax.random.key(33)})
```
Here too.

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 12]
```diff
@@ -0,0 +1,774 @@
...
+    hash_int = produce_hash(data=initial_data+(call_count,))
+    rng_key = jax.random.fold_in(stream_seed, jnp.uint32(hash_int))
+    print(f"{module_name}, count {call_count}: {rng_key}")
+```
+
+### Using `self.param` and `self.variable`
+
++++
+
+Flax users have the option of creating additional parameters and variables in their modules by using the [`self.param`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.param) and [`self.variable`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.variable) `Module` methods. An `init_fn` argument must be passed to these methods so that it can generate the initial value of the parameter / variable. `self.make_rng` is commonly used implicitly or explicitly in this `init_fn`, since many initializer functions are stochastic in nature and require a PRNG key. See the full list of Flax initializers [here](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/initializers.html).
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:277]: Nit:

```suggestion
You can create additional parameters and variables in their `Module`s by using the [`self.param`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.param) and [`self.variable`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/module.html#flax.linen.Module.variable) `Module` methods.
- Remember: You must pass the `init_fn` argument these methods, so that it can generate the initial value of the parameter/variable.
- `self.make_rng` is commonly used implicitly or explicitly in the `init_fn` argument, since many initializer functions are stochastic in nature and require a PRNG key.

For a full list of Flax initializers, check out [`flax.linen.initializers`](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/initializers.html).
```

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 13]
```diff
@@ -0,0 +1,774 @@
...
+  ((), 1, other_seed, (2, 2)),
+  (('Dense_1',), 1, params_seed, (2, 2)),
+  (('Dense_1',), 2, params_seed, (1, 2)),
+):
+  hash_int = produce_hash(data=(*initial_data, count))
+  rng_key = jax.random.fold_in(seed, jnp.uint32(hash_int))
+  print(jax.random.normal(rng_key, shape))
+```
+
+### Managing RNG streams in a training loop
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:348]:
```suggestion
### Managing PRNG streams inside a training loop
```

**Review by 8bitmp3 — COMMENTED** [review]
A few more comments, LMKWYT.

**Review by 8bitmp3 — APPROVED** [review]
thanks

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 14]
```diff
@@ -0,0 +1,797 @@
...
+:outputId: 28bfe59b-ded1-455a-c894-f1c457ec28bf
+
+jax.debug.visualize_array_sharding(out)
+```
+
+## Lifted transforms
+
++++
+
+[Flax lifted transforms](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/transformations.html) allow you to use [jax transforms](https://github.com/google/jax#transformations) with `Module` arguments. This section will show you how to control how PRNG keys are split in Flax lifted transforms.
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:651]:
```suggestion
[Flax lifted transforms](https://flax.readthedocs.io/en/latest/api_reference/flax.linen/transformations.html) allow you to use [JAX transforms](https://github.com/google/jax#transformations) with `Module` arguments. This section will show you how to control how PRNG keys are split in the Flax lifted transforms.
```

File: docs/guides/flax_fundamentals/rng_guide.md [hunk 15]
```diff
@@ -0,0 +1,797 @@
...
+import hashlib
+```
+
+```{code-cell}
+:outputId: ab5e0d3b-2d51-44ee-b823-d152ad1a10b2
+
+jax.devices()
+```
+
+Set Jax config variable `jax_threefry_partitionable` to `True`. This will be the default value in the future and makes the RNG more efficiently auto-parallelizable under `jax.jit`. See this [JAX discussion](https://github.com/google/jax/discussions/18480) for more details.
```
Comment (8bitmp3) [inline docs/guides/flax_fundamentals/rng_guide.md:59]: Nit:
```suggestion
Set the JAX config variable `jax_threefry_partitionable` to `True`. This will be the default value in the future and makes the PRNG more efficiently auto-parallelizable under `jax.jit`. Refre to this [JAX discussion](https://github.com/google/jax/discussions/18480) for more details.
```

**Review by 8bitmp3 — APPROVED** [review]
thank you @chiamp
